jwst/outlier_detection/utils.py

```diff
@@ -20,177 +26,365 @@
+        # Save median model to fits
+        median_model = datamodels.ImageModel(median_data)
+        median_model.update(drizzled_model)
+        median_model.meta.wcs = median_wcs
+        save_median(median_model, make_output_path, resamp.asn_id)
+        del median_model
+    del drizzled_model
 
+    return median_data, median_wcs
 
-def _get_sections(library, nsections, section_nrows, imrows):
-    """Iterator to return sections from a ModelLibrary.
 
-    Parameters
-    ----------
-    library : ModelLibrary
-        The input data models.
+class DiskAppendableArray:
+    """
+    Creates a temporary file to which to append data, in order to perform
+    timewise operations on a stack of input images without holding all of them
+    in memory.
+    
+    This class is purpose-built for the median computation during outlier detection 
+    and is not very flexible. It is assumed that each data array passed to `append`
+    represents the same spatial segment of the full dataset. It is also assumed that 
+    each data array passed to `append` represents only a single instant in time; 
+    the append operation will stack them along a new axis.
+    
+    The `read` operation is only capable of loading the full array back into memory.
+    When working with large datasets that do not fit in memory, the
+    required workflow is to create many DiskAppendableArray objects, each holding
+    a small spatial segment of the full dataset.
+    """
+    
+    def __init__(self, slice_shape, dtype, filename):
+        """
+        Parameters
+        ----------
+        slice_shape : tuple
+            The shape of the spatial section of input data to be appended to the array.
+
+        dtype : str
+            The data type of the array. Must be a valid numpy array datatype.
+
+        filename : str
+            The full file path in which to store the array
+        """
+        if len(slice_shape) != 2:
+            raise ValueError(f"Invalid slice_shape {slice_shape}. Only 2-D arrays are supported.")
+        self._filename = Path(filename)
+        with open(filename, "wb") as f:   # noqa: F841
+            pass
+        self._slice_shape = slice_shape
+        self._dtype = np.dtype(dtype)
+        self._append_count = 0
+
+
+    @property
+    def shape(self):
+        return (self._append_count,) + self._slice_shape
+
+
+    def append(self, data):
+        """Add a new slice to the temporary file"""
+        if data.shape != self._slice_shape:
+            raise ValueError(f"Data shape {data.shape} does not match slice shape {self._slice_shape}")
+        if data.dtype != self._dtype:
+            raise ValueError(f"Data dtype {data.dtype} does not match array dtype {self._dtype}")
+        with open(self._filename, "ab") as f:
+            data.tofile(f, sep="")
+            #f.write(data.tobytes())
+        self._append_count += 1
+
+
+    def read(self):
+        """Read the 3-D array into memory, then delete the tempfile"""
```

Comment:
I'm failing to see where the file is deleted in this method. Is this docstring out-of-date?

Reply:
yes it's out of date